src/lib/storage/lz4_segment.hpp

```diff
@@ -0,0 +1,85 @@
+#include <memory>
+
+#include "base_encoded_segment.hpp"
+#include "storage/vector_compression/base_compressed_vector.hpp"
+#include "types.hpp"
+
+namespace opossum {
+
+class BaseCompressedVector;
+
+template <typename T>
+class LZ4Segment : public BaseEncodedSegment {
+ public:
+  /*
+   * An LZ4 segment is a container for an LZ4 compressed segment. It contains the compressed data, the necessary
+   * metadata and the ability to decompress the data again.
+   *
+   * @param compressed_data The char vector that contains the LZ4 compressed segment data as binary blob.
+   * @param null_values Boolean vector that contains the information which row is null and which is not null.
+   * @param offsets If this segment is not a std::string segment this will be a null pointer. Otherwise it contains
+   *                the offsets for the compressed strings. The offset at position 0 is the character index of the
+   *                string at index 0. Its (exclusive) end is at the offset at position 1. The last string ends at the
+   *                end of the compressed data (since there is offset after it that specifies the end offset).
+   * @param compressed_size The size of the compressed data vector (the return value of LZ4)
+   * @param decompressed_size The size in bytes of the decompressed data vector.
+   * @param num_elements The number of rows that are compressed in this segment.
+   */
+  explicit LZ4Segment(const std::shared_ptr<const pmr_vector<char>>& compressed_data,
+                      const std::shared_ptr<const pmr_vector<bool>>& null_values,
+                      const std::shared_ptr<const pmr_vector<size_t>>& offsets, const int compressed_size,
+                      const int decompressed_size, const size_t num_elements);
+
+  std::shared_ptr<const pmr_vector<char>> compressed_data() const;
+  std::shared_ptr<const pmr_vector<bool>> null_values() const;
+  std::shared_ptr<const pmr_vector<size_t>> offsets() const;
+  int decompressed_size() const;
+
+  /**
+   * @defgroup BaseSegment interface
+   * @{
+   */
+
+  const AllTypeVariant operator[](const ChunkOffset chunk_offset) const final;
+
+  const std::optional<T> get_typed_value(const ChunkOffset chunk_offset) const;
+
+  size_t size() const final;
+
+  std::shared_ptr<std::vector<T>> decompress() const;
+
+  std::shared_ptr<BaseSegment> copy_using_allocator(const PolymorphicAllocator<size_t>& alloc) const final;
+
+  size_t estimate_memory_usage() const final;
+
+  /**@}*/
+
+  /**
+   * @defgroup BaseEncodedSegment interface
+   * @{
+   */
+
+  EncodingType encoding_type() const final;
+  std::optional<CompressedVectorType> compressed_vector_type() const final;
+
+  /**@}*/
+
+ private:
+  const std::shared_ptr<const pmr_vector<char>> _compressed_data;
+  const std::shared_ptr<const pmr_vector<bool>> _null_values;
+  const std::shared_ptr<const pmr_vector<size_t>> _offsets;
+  const int _compressed_size;
+  const int _decompressed_size;
+  size_t _num_elements;
```

Comment:
Why isn't this const?

Reply:
We deleted it (since null_values.size is the same).